utils.c

```diff
@@ -883,3 +884,82 @@ label_exec (UNUSED const char *exec_label)
 #endif
   return 0;
 }
+
+void
+strappend (StringBuilder *dest, const char *src)
+{
+  size_t len = strlen (src);
+
+  if (dest->offset + len >= dest->size)
+    {
+      dest->size = (dest->size + len + 1) * 2;
+      dest->str = xrealloc (dest->str, dest->size);
+    }
+
+  strncpy (dest->str + dest->offset, src, len + 1);
+  dest->offset += len;
+}
+
+__attribute__((format (printf, 2, 3)))
+void
+strappendf (StringBuilder *dest, const char *fmt, ...)
+{
+  va_list args;
+  int len;
+
+  va_start (args, fmt);
+  len = vsnprintf (dest->str + dest->offset, dest->size - dest->offset, fmt, args);
+  va_end (args);
+  if (len < 0)
+    die_with_error ("vsnprintf");
+  if (dest->offset + len >= dest->size)
+    {
+      dest->size = (dest->size + len + 1) * 2;
+      dest->str = xrealloc (dest->str, dest->size);
+      va_start (args, fmt);
+      len = vsnprintf (dest->str + dest->offset, dest->size - dest->offset, fmt, args);
+      va_end (args);
+      if (len < 0)
```

Comment:
could even add
```suggestion
      if (len < 0 || dest->offset + len >= dest->size)
```

Reply:
One could, but why? The added test will always be true if nothing mucks around with the internals. This `die_with_error` is only here in case `vsnprintf` blows up.